File: Solution/0133.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* DFS(Node* node, unordered_map<Node*, Node*>& visited){
        Node* copyNode = new Node(node->val);

        visited.insert({node, copyNode});

        for(int i=0 ; i<node->neighbors.size(); ++i){
            if(visited.find(node->neighbors[i]) != visited.end())
                copyNode->neighbors.push_back(visited[node->neighbors[i]]);
            else{
                Node* newNode = DFS(node->neighbors[i], visited);
                copyNode->neighbors.push_back(newNode);
            }
        }
        return copyNode;
    }

    Node* cloneGraph(Node* node) {
        if(!node) return NULL;
        unordered_map<Node*, Node*> visited;

        return DFS(node, visited);
    }
};

File: Solution/0133.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def DFS(self, node):
        if(node in self.visitedHash.keys()):
            return self.visitedHash[node];
        
        clonedNode = Node(node.val)
        self.visitedHash[node] = clonedNode
        for neighbor in node.neighbors:
            clonedNode.neighbors.append(self.DFS(neighbor))
        return self.visitedHash[node]
    
    def cloneGraph(self, node: 'Node') -> 'Node':
        if(node == None): 
            return node
        self.visitedHash = {}
        
        return self.DFS(node)
        